### PR TITLE
Add core functionality for parsing attributes from an open file

### PR DESCRIPTION
**Pull request by andersy005**
This PR adds a `Builder` class. This class implements most of the functionality needed to parse attributes from an open file/xarray dataset. The missing feature consists of parsing attributes from a filename/filepath template + global attributes i.e. @sherimickelson's work in https://github.com/NCAR/intake-esm-datastore/pull/70. 

@sherimickelson, I added [sample files](https://github.com/NCAR/ecg/tree/master/sample_data) to this repo. I am hoping that we can use them to test your work in https://github.com/NCAR/intake-esm-datastore/pull/70 once it's migrated here.

### CODE REVIEW

**Comment by andersy005** [comment]
Here's a [demo notebook](https://nbviewer.jupyter.org/gist/andersy005/d18f16e744f83c456be7f06312890740)

File: ecgtools/core.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        """
+        Create catalog as a Pandas DataFrame.
+
+        Returns
+        -------
+        `ecgtools.Builder`
+
+        """
+        self.df = pd.DataFrame(self.entries)
+        return self
```
[the remark this answers is not in the document]
Comment (andersy005) [reply ecgtools/core.py:243]: > The user has a root_path (inside of which are files for which they want to generate a catalog). They create a "mapping" function (like the cmip6_parser in the tutorial), and then they call a function to create the catalog. Hence, the result should be a json file and a matching csv file.

👍 I concur. I made the `get_*` methods private, and simplified the API. Here's what I have so far:

```python

builder = ecgtools.Builder(
    root_path="../../sample_data/cmip/CMIP6/",
    extension="*.nc",
    depth=4,
    exclude_patterns=["*/files/*", "*/latest/*"],
)

builder.parse_files_attributes(global_attrs, parser=cmip6_parser, lazy=False)\
  .create_catalog(path="my_catalog.csv")
```


Regarding the JSON creation, I am still looking into it.

**Comment by andersy005** [comment]
Note: the latest commit includes the necessary code for creating both the CSV and JSON files.

File: ecgtools/core.py [hunk 2]
```diff
@@ -168,24 +171,24 @@ def get_filelist_from_dirs(
...
             ]
         else:
             filelist = [
-                self.get_filelist(directory, extension, exclude_patterns) for directory in dirs
+                self._get_filelist(directory, extension, exclude_patterns) for directory in dirs
             ]
         self.filelist = filelist
         return self
 
     def parse_files_attributes(
```
Comment (kmpaul) [inline ecgtools/core.py:190]: Specifically on the `parse_files_attributes` method...

This seems like a helper function.  The usefulness to the user (see previous comment) seems to be the creation of the `entries` attribute.  But, in its current form, it seems like `entries` is less useful than `df` which is just the `entries` attribute converted to a DataFrame.  Correct?  Maybe, if this function should remain public, the end product of this function is the generation of the `df` attribute, and then make `entries` private.  If so, then maybe the function should change its name to reflect that.

Also, this method calls a public function of the same name that exists at module scope (below).  This seems very confusing.  If the two functions did exactly the same thing, then perhaps it makes sense for the functions to have the same name, but they do not do the same thing, exactly.

So, I might pose the following solutions:

- Is the module-scope public function `parse_files_attributes` really something you want in the public API?  Or is it really more of a helper function?  If so, maybe make it private.
- If `Builder.parse_files_attributes` is most useful as a step to constructing the DataFrame, then maybe change the function's name to match that capability (like `Builder.create_dataframe`)?

File: ecgtools/core.py [hunk 3]
```diff
@@ -213,21 +216,125 @@ def parse_files_attributes(
...
         `ecgtools.Builder`
 
+        Notes
+        -----
+        See https://github.com/NCAR/esm-collection-spec/blob/master/collection-spec/collection-spec.md
+        for more details on ESM collection specification.
+
         """
+
         self.df = pd.DataFrame(self.entries)
```
Comment (kmpaul) [inline ecgtools/core.py:291]: If you want this to be publicly available, then I recommend moving this to the `Builder.parse_files_attributes` function (and renaming the function).  And make `self.entries` private.

File: ecgtools/core.py [hunk 4]
```diff
@@ -213,21 +216,125 @@ def parse_files_attributes(
...
+        esmcol_data['aggregation_control'] = {
+            'variable_column_name': variable_column,
+            'groupby_attrs': groupby_attrs,
+            'aggregations': aggregations,
+        }
+
+        json_path = f'{catalog_file.split(".")[0]}.json'
+
+        with open(json_path, mode='w') as outfile:
+            json.dump(esmcol_data, outfile, indent=2)
```
Comment (kmpaul) [inline ecgtools/core.py:334]: Should the `esmcol_data` object be made public?  There is a public version of the CSV file in the `df` attribute, so maybe a public version of the JSON file should be made public?

See my notes in the review summary.

**Comment by andersy005** [comment]
I like the simplicity of the public API suggested above, but I do have a few questions:

> Builder.__init__(): The result of constructing the Builder should be a valid object that knows (as well as possible) that it will run without error. 


Could you expand on this? What errors should we check for? I am asking because I think this can be very tricky. For example, if the user provides a `root_path`, do we check to see that they have the right permissions to crawl the sub-directories during the class instantiation step? 


> Builder.build(): The result of the Builder.build method should be the construction of the Builder.df and (possibly) Builder.dict (?) attributes, described below:

Does this imply that all necessary arguments are handled in `__init__()`?


> Does that cover the user behavior you are aiming for?

One of the things I was aiming for was a Builder object that behaves as a pipeline i.e. one that makes it easy to replace intermediate steps if need be. For example, when building a large catalog, for instance CMIP6 catalog, you may want to run the pipeline end-to-end for the first time only. For the subsequent runs, you would only collect the list of files available, compare this list to the list of files in the existing catalog, and only parse attributes of files that don't appear in the catalog previously built. My understanding of the public API you suggested above is that all steps are tightly coupled, making it a little tricky to replace intermediate steps when necessary. I could be missing something though.

File: ecgtools/core.py (hunk 2 above)
Comment (andersy005) [reply ecgtools/core.py:190]: > But, in its current form, it seems like entries is less useful than df which is just the entries attribute converted to a DataFrame. Correct? 

I concur. `entries` is less useful than df. The most important pieces here are the `filelist` and the `df`. 

> Is the module-scope public function parse_files_attributes really something you want in the public API? Or is it really more of a helper function? If so, maybe make it private.

👍Will make it private, and move its arguments into `__init__()`

**Comment by kmpaul** [comment]
> @kmpaul, I like the simplicity of the public API suggested above, but I do have a few questions:
> 
> > `Builder.__init__()`: The result of constructing the `Builder` should be a valid object that knows (as well as possible) that it will run without error.
> 
> Could you expand on "it will run without error"? What errors should we check for? I am asking because I think this can be very tricky. For example, if the user provides a `root_path`, do we check whether they have the right permissions to crawl the sub-directories during the class instantiation step?

That's one idea, or you can just check to make sure all of the necessary inputs to the `__init__()` constructor are valid.  If a `root_path` is given, then it makes sense to make sure that path is valid and you have access to it, but maybe you don't have to crawl the entire tree.  Just make sure the `__init__()` inputs are valid.  If the validity checks are too complex, then you should just throw an exception upon error when the operation runs.

Then you need validation unit tests to match.

> > `Builder.build()`: The result of the `Builder.build` method should be the construction of the `Builder.df` and (possibly) `Builder.dict` (?) attributes, described below:
> 
> Does this imply that all necessary arguments are handled in `__init__()`?

Not necessarily.  The `build()` method could take arguments of its own, as long as it makes sense for them to be supplied to `build()` instead of `__init__()`, right?  From my understanding, the `build` function is really just the `parse_files_attributes` method *and* the current `create_catalog` method, combined.  So, as a first pass, you could make the `parse_files_attributes` and `create_catalog` methods private and called in sequence from a `build` method.  Then do a little cleanup.

> > Does that cover the user behavior you are aiming for?
> 
> One of the things I was aiming for was a Builder object that behaves as a pipeline i.e. one that makes it easy to replace intermediate steps if need be. For example, when building a large catalog, for instance CMIP6 catalog, you may want to run the pipeline end-to-end for the first time only. For the subsequent runs, you would only collect the list of files available, compare this list to the list of files in the existing catalog, and only parse attributes of files that don't appear in the catalog previously built. My understanding of the public API you suggested above is that all steps are tightly coupled, making it a little tricky to replace intermediate steps when necessary. I could be missing something though.

Yes.  I didn't realize this was a use case, but if that use case is to be assumed (and it should), then you are correct.  My pitch is tightly coupled.

So, could you add an `update` method to the public API that addresses your additional use case?  And then modify the private helper functions to allow for the `update` step?

I am envisioning a 2-option use path:

    [init] --> [build] --> [save]
    [init] --> [update] --> [save]

Does that work?  Is that clearer?

**Comment by andersy005** [comment]
It makes sense now :) Many thanks for the feedback

**Comment by kmpaul** [comment]
You're welcome.  I hope what I'm suggesting is reasonable.  I think that if what you are suggesting is the ability to let users **build** an entire catalog from scratch *and* to let users **update** an existing catalog, then it makes sense to call out those 2 scenarios in the names of the API functions.  It will be more clear to the user.

**Comment by andersy005** [comment]
Gentle ping @kmpaul. When you get a chance, this is ready for another round of review.

File: ecgtools/core.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+
+        if aggregations is None:
+            aggregations = []
+        esmcol_data['aggregation_control'] = {
+            'variable_column_name': variable_column,
+            'groupby_attrs': groupby_attrs,
+            'aggregations': aggregations,
+        }
+
+        self.esmcol_data = esmcol_data
```
Comment (kmpaul) [inline ecgtools/core.py:351]: This all seems like a `build` step...?

Comment (andersy005) [reply ecgtools/core.py]: Fixed it

**Comment by andersy005** [comment]
> @andersy005 I like this. I think this is a good public API. I'm happy seeing this merged into the master branch.

Hooray 🎉 @kmpaul & @sherimickelson thank you for your feedback!

**Pull request merged**